File: android/tutorials/tutorial3-completed/src/main/java/com/uber/rib/root/loggedin/LoggedInInteractor.java
/*
 * Copyright (C) 2017. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.rib.root.loggedin;

import androidx.annotation.Nullable;
import com.uber.rib.core.Bundle;
import com.uber.rib.core.EmptyPresenter;
import com.uber.rib.core.Interactor;
import com.uber.rib.core.RibInteractor;
import com.uber.rib.root.loggedin.offgame.OffGameInteractor;
import com.uber.rib.root.loggedin.tictactoe.TicTacToeInteractor;
import javax.inject.Inject;

/** Coordinates Business Logic for {@link LoggedInScope}. */
@RibInteractor
public class LoggedInInteractor extends Interactor<EmptyPresenter, LoggedInRouter> {

  @Inject @LoggedInBuilder.LoggedInInternal MutableScoreStream scoreStream;

  @Override
  protected void didBecomeActive(@Nullable Bundle savedInstanceState) {
    super.didBecomeActive(savedInstanceState);

    // when first logging in we should be in the OffGame state
    getRouter().attachOffGame();
  }

  @Override
  protected void willResignActive() {
    super.willResignActive();

    // TODO: Perform any required clean up here, or delete this method entirely if not needed.
  }

  class OffGameListener implements OffGameInteractor.Listener {

    @Override
    public void onStartGame() {
      getRouter().detachOffGame();
      getRouter().attachTicTacToe();
    }
  }

  class TicTacToeListener implements TicTacToeInteractor.Listener {

    @Override
    public void gameWon(@Nullable String winner) {
      if (winner != null) {
        scoreStream.addVictory(winner);
      }

      getRouter().detachTicTacToe();
      getRouter().attachOffGame();
    }
  }
}

File: android/demos/compose/src/main/kotlin/com/uber/rib/compose/root/main/logged_in/LoggedInInteractor.kt
/*
 * Copyright (C) 2021. Uber Technologies
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.uber.rib.compose.root.main.logged_in

import com.uber.rib.compose.root.main.AuthInfo
import com.uber.rib.compose.root.main.AuthStream
import com.uber.rib.compose.root.main.logged_in.off_game.OffGameInteractor
import com.uber.rib.compose.root.main.logged_in.tic_tac_toe.TicTacToeInteractor
import com.uber.rib.compose.util.EventStream
import com.uber.rib.core.BasicInteractor
import com.uber.rib.core.Bundle
import com.uber.rib.core.ComposePresenter
import com.uber.rib.core.coroutineScope
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class LoggedInInteractor(
  presenter: ComposePresenter,
  private val authInfo: AuthInfo,
  private val authStream: AuthStream,
  private val eventStream: EventStream<LoggedInEvent>,
  private val scoreStream: ScoreStream
) : BasicInteractor<ComposePresenter, LoggedInRouter>(presenter),
  OffGameInteractor.Listener,
  TicTacToeInteractor.Listener {

  override fun didBecomeActive(savedInstanceState: Bundle?) {
    super.didBecomeActive(savedInstanceState)

    eventStream.observe()
      .onEach {
        when (it) {
          is LoggedInEvent.LogOutClick -> authStream.accept(AuthInfo(false))
        }
      }
      .launchIn(coroutineScope)

    router.attachOffGame(authInfo)
  }

  override fun onStartGame() {
    router.detachOffGame()
    router.attachTicTacToe(authInfo)
  }

  override fun onGameWon(winner: String?) {
    if (winner != null) {
      coroutineScope.launch { scoreStream.addVictory(winner) }
    }

    router.detachTicTacToe()
    router.attachOffGame(authInfo)
  }
}
